src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java

```diff
@@ -1218,28 +1264,34 @@ else if (!authCheckService.isAtLeastTeachingAssistantForExercise(programmingExer
     public ResponseEntity<TextPlagiarismResult> checkPlagiarism(@PathVariable long exerciseId, @RequestParam float similarityThreshold, @RequestParam int minimumScore)
             throws ExitException, IOException {
         log.debug("REST request to check plagiarism for ProgrammingExercise with id: {}", exerciseId);
-        long start = System.nanoTime();
 
-        Optional<ProgrammingExercise> programmingExercise = programmingExerciseRepository.findById(exerciseId);
-        if (programmingExercise.isEmpty()) {
+        Optional<ProgrammingExercise> optionalProgrammingExercise = programmingExerciseRepository.findById(exerciseId);
+
+        if (optionalProgrammingExercise.isEmpty()) {
             return notFound();
         }
-        if (!authCheckService.isAtLeastInstructorForExercise(programmingExercise.get())) {
+
+        ProgrammingExercise programmingExercise = optionalProgrammingExercise.get();
+
+        if (!authCheckService.isAtLeastInstructorForExercise(programmingExercise)) {
             return forbidden();
         }
 
-        var language = programmingExercise.get().getProgrammingLanguage();
+        ProgrammingLanguage language = programmingExercise.getProgrammingLanguage();
         ProgrammingLanguageFeature programmingLanguageFeature = programmingLanguageFeatureService.getProgrammingLanguageFeatures(language);
 
         if (!programmingLanguageFeature.isPlagiarismCheckSupported()) {
             return ResponseEntity.badRequest().headers(HeaderUtil.createFailureAlert(applicationName, true, ENTITY_NAME, "programmingLanguageNotSupported",
                     "Artemis does not support plagiarism checks for the programming language " + language)).body(null);
         }
 
+        Optional<PlagiarismResult> optionalPreviousResult = plagiarismService.getPlagiarismResult(programmingExercise);
+
         TextPlagiarismResult result = programmingExerciseExportService.checkPlagiarism(exerciseId, similarityThreshold, minimumScore);
 
-        log.info("Check plagiarism of programming exercise {} with title '{}' was successful in {}.", programmingExercise.get().getId(), programmingExercise.get().getTitle(),
-                formatDurationFrom(start));
+        plagiarismService.savePlagiarismResult(result);
+
+        optionalPreviousResult.ifPresent(plagiarismService::deletePlagiarismResult);
```

Comment:
this logic of saving the new result and deleting the old could be reused by moving it into the plagiarismService.

Reply:
good point